rpc/codec/server.go

```diff
@@ -79,6 +91,58 @@ func (c *serverCodec) ReadRequestHeader(r *rpc.Request) error {
 	return nil
 }
 
+// authenticateRequest takes a request proto and attempts to authenticate it.
+// For requests with a RequestHeader, we compare the header.User against
+// the client certificate Subject.CommonName.
+// For other requests, we either allow all (test requests), or
+// require the "node user".
+func (c *serverCodec) authenticateRequest(request proto.Message) error {
+	var err error
+	var requestedUser string
+
+	switch request := request.(type) {
+	case cockroach_proto.Request:
+		// Request interface: this covers most requests.
+		header := request.Header()
+		if header == nil {
+			return util.Errorf("missing header in request: %+v", request)
+		}
+		requestedUser = header.GetUser()
+		if len(requestedUser) == 0 {
+			return util.Errorf("missing User in request header: %+v", header)
+		}
+	case *cockroach_proto.PingRequest, *cockroach_proto.GossipRequest, *cockroach_proto.RaftMessageRequest:
+		// Ping and Gossip requests do not have a header: require the node user.
+		// TODO(marc): it may be worth it to make them implement Request.
```

Comment:
yup. I started working on that, but ran into protoc issues.
At the very least, I want to remove all test requests.